src/expression.ts

```diff
@@ -157,139 +279,247 @@ export type BinaryOp =
+
   constructor() {
     super("UndefinedLiteralExpr");
   }
+
+  public clone(): this {
+    return new UndefinedLiteralExpr() as this;
+  }
 }
 
 export const isBooleanLiteral = typeGuard("BooleanLiteralExpr");
 
-export class BooleanLiteralExpr extends BaseNode<"BooleanLiteralExpr"> {
+export class BooleanLiteralExpr extends BaseExpr<"BooleanLiteralExpr"> {
   constructor(readonly value: boolean) {
     super("BooleanLiteralExpr");
   }
+
+  public clone(): this {
+    return new BooleanLiteralExpr(this.value) as this;
+  }
 }
 
 export const isNumberLiteralExpr = typeGuard("NumberLiteralExpr");
 
-export class NumberLiteralExpr extends BaseNode<"NumberLiteralExpr"> {
+export class NumberLiteralExpr extends BaseExpr<"NumberLiteralExpr"> {
   constructor(readonly value: number) {
     super("NumberLiteralExpr");
   }
+
+  public clone(): this {
+    return new NumberLiteralExpr(this.value) as this;
+  }
 }
 
 export const isStringLiteralExpr = typeGuard("StringLiteralExpr");
 
-export class StringLiteralExpr extends BaseNode<"StringLiteralExpr"> {
+export class StringLiteralExpr extends BaseExpr<"StringLiteralExpr"> {
   constructor(readonly value: string) {
     super("StringLiteralExpr");
   }
+
+  public clone(): this {
+    return new StringLiteralExpr(this.value) as this;
+  }
 }
 
 export const isArrayLiteralExpr = typeGuard("ArrayLiteralExpr");
 
-export class ArrayLiteralExpr extends BaseNode<"ArrayLiteralExpr"> {
+export class ArrayLiteralExpr extends BaseExpr<"ArrayLiteralExpr"> {
   constructor(readonly items: Expr[]) {
     super("ArrayLiteralExpr");
-    setParent(this, items);
+    items.forEach((item) => item.setParent(this));
+  }
+
+  public clone(): this {
+    return new ArrayLiteralExpr(this.items.map((item) => item.clone())) as this;
   }
 }
 
 export type ObjectElementExpr = PropAssignExpr | SpreadAssignExpr;
 
+export const isObjectElementExpr = typeGuard(
+  "PropAssignExpr",
+  "SpreadAssignExpr"
+);
+
 export const isObjectLiteralExpr = typeGuard("ObjectLiteralExpr");
 
-export class ObjectLiteralExpr extends BaseNode<"ObjectLiteralExpr"> {
+export class ObjectLiteralExpr extends BaseExpr<"ObjectLiteralExpr"> {
   constructor(readonly properties: ObjectElementExpr[]) {
     super("ObjectLiteralExpr");
-    setParent(this, properties);
+    properties.forEach((prop) => prop.setParent(this));
+  }
+
+  public clone(): this {
+    return new ObjectLiteralExpr(
+      this.properties.map((prop) => prop.clone())
+    ) as this;
+  }
+  public getProperty(name: string) {
+    return this.properties.find((prop) => {
+      if (prop.kind === "PropAssignExpr") {
+        if (prop.name.kind === "Identifier") {
+          return prop.name.name === name;
+        } else if (prop.name.kind === "StringLiteralExpr") {
+          return prop.name.value === name;
+        } else if (prop.name.expr.kind === "StringLiteralExpr") {
+          return prop.name.expr.value === name;
+        }
+      }
+      return false;
+    });
   }
```

Comment:
I just removed this because it doesn't seem reliable. for example


```ts
const obj = { val: "val" };
const obj2 = {
   val: "foo",
  [obj.val]: "bar",
   ...obj
}
```

what does `getPropety("val")` return?

Reply:
If we want want to support getting a value from an object...

The object needs to be completely resolvable (eventually constant) or the target language needs to support runtime resolution.

Reply:
It is not reliable - that's true. I only need it for grabbing properties where the names are known.

Reply:
but the result could still be wrong when there is a computed or a spread. Even if you try to resolve the spread and computed, they result is only valid when all are constants.

Reply:
I'd rather it to partially work than not exist. I use it to grab properties like TableName which is a reference to a table. I need it.

Reply:
yeah, I get it. In #39, I try to resolve keys and if they are not constants, I fail. This means computed and spread will work as long as the sources are constant eventually.

You have the same requirement here, but the current impl will fail or be wrong even the spread/computed is constant.

https://github.com/sam-goodwin/functionless/pull/39/files/5087c05d71f2d079e5a40c013cbbb69b1150a1ee#diff-9a722b39e6943df7d0583e4d2083f5fc65e75fa508310c45252774e3c47574c5R189

Flatten turns spread and computed into constants if possible.

https://github.com/sam-goodwin/functionless/pull/39/files/5087c05d71f2d079e5a40c013cbbb69b1150a1ee#diff-9a722b39e6943df7d0583e4d2083f5fc65e75fa508310c45252774e3c47574c5R246

Reply:
I feel like we should either try to resolve constants, or fail if there is any chance that a dynamic prop assignment could contain the key.

```
{
   TableName: default,
   ... props // could contain another table name
}
```

Reply:
Agree with this analysis - you're right, this could be improved. For now, it works for most cases. Will track an improvement in an issue https://github.com/sam-goodwin/functionless/issues/82